WIP: Allow extending the expiry of a reservation

serboctor opened the pull request:
Resolves https://github.com/threefoldtech/grid_broker/issues/13
relevant PR https://github.com/threefoldtech/jumpscaleX/pull/649

serboctor committed:
Allow extending the expiry of a reservation

serboctor committed:
Add expiry date to the first reservation email

zaibon commented on templates/grid_broker/grid_broker.py:
```diff
@@ -62,11 +63,32 @@ def _watch_transactions(self):
+                    action_type = "extension"
+                    title = "Extending reservation failed"
+
+                    expiry_date = self._extend_reservation(tx, data)
```
why does `_extend_reservation` not return the datetime object directly instead of doing the conversion in the next line ?

zaibon commented on templates/grid_broker/grid_broker.py:
```diff
@@ -347,19 +377,30 @@ def _is_locked(self, tx):
+_extend_template = """
+<html>
+<body>
+    <h1>Your reservation {tx_id} has been extending successfully</h1>
```
can we also specify what kind of reservation has been extended ?
Only the transaction ID is not really explicit.

zaibon commented on templates/reservation/reservation.py:
```diff
@@ -30,9 +32,22 @@ def __init__(self, name, guid=None, data=None):
         if not self.data.get('creationTimestamp'):
             self.data['creationTimestamp'] = time.time()
+        if not self.data.get('expiryTimestamp'):
+            self.data['expiryTimestamp'] = time.time()
```
does that means all previous transaction will be marked as expired after we apply this patch ?

serboctor replied:
well, in case of install, I am depending on install setting the extending it. However, I dont like the fact that we don't set the creationtimestamp in the data before creating the service to begin with. I'll add both to the data of the service before creation in the grid_broker. If I remove it those two lines from the validate, should I then raise an error if someone tries to extend a service where the expirytimestamp isnt set? this should theoritically not happen unless someone ceates the service themselves and not through grid_broker.

serboctor replied:
Ok since I now set it in the data before creating a service, I changed the validate to raise an error if either aren't set. I think this makes more sense.

zaibon commented on templates/reservation/reservation.py:
```diff
@@ -30,9 +32,22 @@ def __init__(self, name, guid=None, data=None):
+        if not self.data.get('expiryTimestamp'):
+            self.data['expiryTimestamp'] = time.time()
+
+    def extend(self, duration):
```
This is a bit hard to review like that. Can you add a bunch of test to validate the calculation of the extended time is correct ?

serboctor replied:
I moved the functionality under `j.tools.time` and added a test there.
https://github.com/threefoldtech/jumpscale_lib/commit/12ec507e86ab2d0e8593ac64c5a8004d0b018642

zaibon commented on templates/reservation/reservation.py:
```diff
@@ -30,9 +32,22 @@ def __init__(self, name, guid=None, data=None):
 
     @retry((Exception), tries=4, delay=5, backoff=2, logger=None)
-    def install(self):
+    def install(self,duration):
```
This is very unusuall to have a argument with the `install` action, why is this required ?

serboctor replied:
Yea, actually I can just compute it in the grid_broker and set it in the data.

zaibon replied:
don't forget to remove it

zaibon commented on templates/reservation/reservation.py:
```diff
@@ -50,8 +65,9 @@ def install(self):
+        self.extend(duration)
         self.state.set('actions', 'install', 'ok')
-        return install_result
+        return install_result, self.data["expiryTimestamp"]
```
I don't think you can return 2 values from an action.
I think this will create an undefined behavior on the client side that doesn't not how to deal with this value.

serboctor committed:
Refactor expiryTimestamp flow

serboctor committed:
Fix typo in email template

serboctor committed:
Remove unnecessary import

serboctor committed:
Check the bot expiration before extending reservation

serboctor committed:
Fix reservation typo

zaibon commented on templates/reservation/reservation.py:
```diff
@@ -28,11 +27,23 @@ def __init__(self, name, guid=None, data=None):
     def validate(self):
-        if not self.data.get('creationTimestamp'):
-            self.data['creationTimestamp'] = time.time()
+        for key in ['creationTimestamp', 'expiryTimestamp']:
```
What about already existing services?
they won't have a value on the expiryTimestamp

serboctor replied:
May be we can have some sort of migration script/function? and for existing active reservations that have a creationtimestamp and no expiry, it sets the expiry to 1 month after the creation? Because I believe the validation needs to be there.

zaibon replied:
> for existing active reservations that have a creationtimestamp and no expiry, it sets the expiry to 1 month after the creation? 

Good idea. So the template can update the data of the services automatically

serboctor committed:
Harden transaction refund

zaibon commented on templates/grid_broker/grid_broker.py:
```diff
@@ -45,53 +48,99 @@ def _watch_transactions(self):
                 # refund if there is not data
                 if not data:
+                    self.logger.info("no data found in transaction %s", tx.id)
+                    raise ValueError("transaction has no data")
```
This is not right. Here you raise cause there is no data.
So you end up in the except block line 58. And in this block you try to refund. But refund will always fail if there is not data cause you don't know the email address of the sender.

zaibon replied:
Also I don't really understand why you need the `malformed` variable. The only thing that happens in this try block is the `_parse_tx_data`.  So the only case where you can end up in the except block is if the parsing of the transaction failed.

zaibon replied:
I think we need to split this more, something like:
```python
data = None
try:
   data = parse_tx(data)
except:
   log.error

if not data:
    log error
     return

try:
    deploy
except:
    refund
```

serboctor replied:
First of all, I didn't change the logic at all, I only restructured it so we don't have a refund block twice and hardened it so there is a try/except around the refund. It already did used to refund if there was no data. I don't understand the relationship between data being empty and not being able to refund. This is the data from the notary, the refund function doesn't use the data anyway. Also, if we can attempt to refund when the data can't be parsed, then we can definitely attempt to refund when the data is empty because when parsing fails, we also end up with no data. I just raise a value error so I can take advantage of the refund in the try/except instead of having duplicate code just two lines above it. The `malformed` variable is just there for sake of logging so I can differentiate between both cases.

zaibon replied:
yeah I got confused, the data is indeed not needed to refund, just to notify the user via email.

serboctor committed:
Refactor reservation:
- Change the 3bot expiry comparison to ignore hour/minute.
- Add more logs in the grid_broker.
- Fix 3bot id usage bug.

serboctor committed:
Add a migration function for old reservation services

serboctor committed:
Fix reservation migration function
